src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1057 @@
+    } else {
+        vcpu_writereg(cpu()->vcpu, acc->reg, vaplic_get_target(cpu()->vcpu, intp));
+    }
+}
+
+static void vaplic_emul_idelivery_access(struct emul_access *acc){
+    int idc_id = ((acc->addr - APLIC_IDC_OFF) >> 5) & 0x3ff;
+    if (acc->write) {
+        vaplic_set_idelivery(cpu()->vcpu, idc_id, vcpu_readreg(cpu()->vcpu, acc->reg));
+    } else {
+        vcpu_writereg(cpu()->vcpu, acc->reg, vaplic_get_idelivery(cpu()->vcpu, idc_id));
+    }
+}
+
+static void vaplic_emul_iforce_access(struct emul_access *acc){
+    int idc_id = ((acc->addr - APLIC_IDC_OFF) >> 5) & 0x3ff;
+    if (acc->write) {
+        vaplic_set_iforce(cpu()->vcpu, idc_id, vcpu_readreg(cpu()->vcpu, acc->reg));
+    } else {
+        vcpu_writereg(cpu()->vcpu, acc->reg, vaplic_get_iforce(cpu()->vcpu, idc_id));
+    }
+}
+
+static void vaplic_emul_ithreshold_access(struct emul_access *acc){
+    int idc_id = ((acc->addr - APLIC_IDC_OFF) >> 5) & 0x3ff;
+    if (acc->write) {
+        vaplic_set_ithreshold(cpu()->vcpu, idc_id, vcpu_readreg(cpu()->vcpu, acc->reg));
+    } else {
+        vcpu_writereg(cpu()->vcpu, acc->reg, vaplic_get_ithreshold(cpu()->vcpu, idc_id));
+    }
+}
+
+static void vaplic_emul_topi_access(struct emul_access *acc){
+    int idc_id = ((acc->addr - APLIC_IDC_OFF) >> 5) & 0x3ff;
+
+    if (acc->write) return;
+    vcpu_writereg(cpu()->vcpu, acc->reg, vaplic_get_topi(cpu()->vcpu, idc_id));
+}
+
+static void vaplic_emul_claimi_access(struct emul_access *acc){
+    int idc_id = ((acc->addr - APLIC_IDC_OFF) >> 5) & 0x3ff;
+
+    if (acc->write) return;
+    vcpu_writereg(cpu()->vcpu, acc->reg, vaplic_get_claimi(cpu()->vcpu, idc_id));
+}
+
+// ============================================================================
+void vaplic_inject(struct vcpu *vcpu, irqid_t intp_id)
+{
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    spin_lock(&virqc->lock);
+    
+    /** Intp has a valid ID and the virtual interrupt is not pending*/
+    if (intp_id > 0 && intp_id < APLIC_MAX_INTERRUPTS && !vaplic_get_pend(vcpu, intp_id)){
+        if(virqc->srccfg[intp_id-1] != APLIC_SOURCECFG_SM_INACTIVE &&
+           virqc->srccfg[intp_id-1] != APLIC_SOURCECFG_SM_DETACH){
+            set_bit_from_reg(&virqc->setip[intp_id/32], intp_id);
+        }
+        vaplic_update_hart_line(vcpu);
+    }
+    spin_unlock(&virqc->lock);
+}
+
+/**
+ * @brief Function to handle writes (or reads) to (from) domain structure.
+ * 
+ * @param acc emulated access
+ * @return true if conclude without errors.
+ * @return false if the access is not aligned.
+ */
+static bool vaplic_domain_emul_handler(struct emul_access *acc)
+{
+    // only allow aligned word accesses
+    if (acc->width != 4 || acc->addr & 0x3) return false;
+
+    switch (acc->addr & 0xffff) {
+        case APLIC_DOMAIN_OFF:
+            vaplic_emul_domaincfg_access(acc);
+            break;
+        case APLIC_SOURCECFG_OFF ... APLIC_SOURCECFG_OFF+((APLIC_MAX_INTERRUPTS-2)*4):
```

Comment:
Since we have a lot of reserved fields in the middle of the domain regmap we could probably do a check of the reserved fields first. If we do it we simplify the logic when we are at case 2. Otherwise, when we want to access a *ip/enum reg we need to do `if (addr == *ip/enum)`.

What I am suggesting is: 
```
static bool vaplic_domain_emul_handler(struct emul_access *acc)
{
    uint16_t emul_addr = 0;

    // only allow aligned word accesses
    if (acc->width != 4 || acc->addr & 0x3) return false;

    emul_addr = acc->addr & 0x3fff;

    if (vaplic_domain_emul_reserved(emul_addr)){
        if(!acc->write) {
            vcpu_writereg(cpu()->vcpu, acc->reg, 0);
        }
        return true;
    }

    switch (emul_addr >> 12)
    {
        case 0:
            if (emul_addr == 0x0000) { /** domaincfg */
                vaplic_emul_domaincfg_access(acc);
            } else { /** sourcecfg */
                vaplic_emul_srccfg_access(acc);
            }
            break;
        case 1:
            switch (emul_addr >> 7)
            {
            case 0x38: /** setip */
                vaplic_emul_setip_access(acc);
                break;
            case 0x39: /** setipnum */
                vaplic_emul_setipnum_access(acc);
                break;
            case 0x3A: /** in_clrip */
                vaplic_emul_in_clrip_access(acc);
                break;
            case 0x3B: /** clripnum */
                vaplic_emul_clripnum_access(acc);
                break;
            case 0x3C: /** setie */
                vaplic_emul_setie_access(acc);
                break;
            case 0x3D: /** setienum */
                vaplic_emul_setienum_access(acc);
                break;
            case 0x3E: /** clrie */
                vaplic_emul_clrie_access(acc);
                break;
            case 0x3F: /** clrienum */
                vaplic_emul_clrienum_access(acc);
                break;
            default:
                if(!acc->write) {
                    vcpu_writereg(cpu()->vcpu, acc->reg, 0);
                }
                break;
            }
            break;
        case 3:
            if (emul_addr == 0x3000) { /** genmsi */
                vaplic_emul_domaincfg_access(acc);
            } else { /** target */
                vaplic_emul_target_access(acc);
            }
            break;
        default:
            if(!acc->write) {
                vcpu_writereg(cpu()->vcpu, acc->reg, 0);
            }
            break;
    }

    return true;
}
```

What are your thoughts on this? Do you think we should examine within each field, or is it possible to adopt this approach?

Reply:
I noticed now that I am not subtracting aplic domain virt base to emul_addr. I already fixed it.

Reply:
I have also updated the hardcoded values to something like this:

```
case offsetof(struct aplic_global_hw, setip) >> 7:
        vaplic_emul_setip_access(ACC);
break;
```
As you suggest in one of your comments.

Reply:
the proposed code adds another return point.... So I create a bool variable called read_only_zero. Thus, when we check is it is reserved we only set this var, ELSE we exec. the case checks. In the end, I added:

```
  if (read_only_zero){
        if(!acc->write) {
            vcpu_writereg(cpu()->vcpu, acc->reg, 0);
        }
    }
```

This allows us to remove the same code from the else branch, making the code cleaner.